lib/proxy_api/resource.rb

```diff
@@ -62,27 +56,72 @@ def parse(response)
 
     # Perform GET operation on the supplied path
     def get(path = nil, payload = {})
-      # This ensures that an extra "/" is not generated
-      if path
-        resource[URI.escape(path)].get payload
-      else
-        resource.get payload
+      with_logger do
+        # This ensures that an extra "/" is not generated
+        if path
+          resource[URI.escape(path)].get payload
+        else
+          resource.get payload
+        end
       end
     end
 
     # Perform POST operation with the supplied payload on the supplied path
     def post(payload, path = "")
-      resource[path].post payload
+      logger.debug("POST request payload: #{payload}")
+      with_logger do
+        resource[path].post payload
+      end
     end
 
     # Perform PUT operation with the supplied payload on the supplied path
     def put(payload, path = "")
-      resource[path].put payload
+      logger.debug("PUT request payload: #{payload}")
+      with_logger do
+        resource[path].put payload
+      end
     end
 
     # Perform DELETE operation on the supplied path
     def delete(path)
-      resource[path].delete
+      with_logger do
+        resource[path].delete
+      end
+    end
+
+    def with_logger
+      old_logger = RestClient.log
+      RestClient.log = RestClientLogger.new(logger)
+      yield
+    ensure
+      RestClient.log = old_logger
+    end
+
+    def request_id
+      ::Logging.mdc['session'] || SecureRandom.uuid
```

Comment:
```SecureRandom.uuid``` returns a random uuid (formatted as such). You could use ```SecureRandom.hex(64)``` to mimic session's digest computed in https://github.com/theforeman/foreman/pull/3683/files#diff-cd38d713fc5c7f078ba5cc991b015720R31.